File: Public/Gui/Linn/Gui/SceneGraph/NodePolygon.py
from Ft.Xml import XPath
import Linn.Gui.Serialize
import Linn.Gui.Renderer
import Linn.Gui.SceneGraph.NodeHit
import Linn.Gui.SceneGraph.RenderState


class NodePolygon(Linn.Gui.SceneGraph.NodeHit.NodeHit):
    def __init__(self, aName=None, aWidth=0, aHeight=0, aTextureScale=False):
        if aName == None:
            aName = self.UniqueName('NodePolygon')
        Linn.Gui.SceneGraph.NodeHit.NodeHit.__init__(self, aName, aWidth, aHeight)
        self.iRenderState = Linn.Gui.SceneGraph.RenderState.RenderState()
        self.iTextureScale = aTextureScale
        
    def Load(self, aStream):
        Linn.Gui.SceneGraph.NodeHit.NodeHit.Load(self, aStream)
        
        rs = aStream.LoadNode('RenderState')
        if len(rs) != 1:
            raise Linn.Gui.Serialize.InvalidLayoutFile('<RenderState>', len(rs), 1)
        if rs[0].hasChildNodes():
            aStream.PushContextNode(rs[0])
            self.iRenderState.Load(aStream)
            aStream.PopContextNode()
        
    def Save(self, aStream):
        if aStream.ContextAttribute('class') == None:
            aStream.SetContextAttribute('class', 'NodePolygon')
            
        Linn.Gui.SceneGraph.NodeHit.NodeHit.Save(self, aStream)
        
        self.iRenderState.Save(aStream)
        
    def Link(self):
        #print 'Linking', self.Name(), '[NodePolygon]'
        Linn.Gui.SceneGraph.NodeHit.NodeHit.Link(self)

        self.iRenderState.Link()
        #self.Texture(self.iRenderState.Texture())
        
    def Update(self, aRecurse=True):
        Linn.Gui.SceneGraph.NodeHit.NodeHit.Update(self, aRecurse)
        if not self.iTextureScale and self.iRenderState.Texture():
            if self.iWidth != self.iRenderState.Texture().Width() or self.iHeight != self.iRenderState.Texture().Height():
                self.iWidth = self.iRenderState.Texture().Width()
                self.iHeight = self.iRenderState.Texture().Height()
                self.ObserverUpdate()
        
    def RenderState(self):
        return self.iRenderState
    
    def SetRenderState(self, aRenderState):
        self.iRenderState = aRenderState
        
    def Draw(self, aRenderer):
        # NOTE: this should not be done here... Some sort of refresh required on texture reload
        if not self.iTextureScale and self.iRenderState.Texture():
            if self.iWidth != self.iRenderState.Texture().Width() and self.iHeight != self.iRenderState.Texture().Height():
                self.iWidth = self.iRenderState.Texture().Width()
                self.iHeight = self.iRenderState.Texture().Height()
                self.ObserverUpdate()
        aRenderer.DrawPolygon(self)
        
    def Visit(self, aVisitor):
        Linn.Gui.SceneGraph.NodeHit.NodeHit.Visit(self, aVisitor)
        # accept visitor for this node
        aVisitor.AcceptPolygon(self)
        
    def Texture(self, aTexture):
        self.iRenderState.SetTexture(aTexture)
        if not self.iTextureScale and aTexture:
            self.iWidth = aTexture.Width()
            self.iHeight = aTexture.Height()
            self.ObserverUpdate()
        
    def ScaleTexture(self):
        return self.iTextureScale





            

File: Public/Gui/Linn/Gui/SceneGraph/NodePolygon.cs
using Linn;
using System.Xml;
using Linn.Gui.Scenegraph;
using Linn.Gui.Resources;
using System;

namespace Linn {
namespace Gui {
namespace Scenegraph {
    
public class NodePolygon : NodeHit
{
    public NodePolygon() : base(UniqueName("NodePolygon")) {
    }
    
    public NodePolygon(string aName) : base(aName) {
    }
    
    public override Object Clone() {
        NodePolygon p = new NodePolygon();
        Clone(p);
        return p;
    }
    
    protected void Clone(NodePolygon aNodePolygon) {
        Clone((NodeHit)aNodePolygon);
        aNodePolygon.iRenderState = iRenderState;
        aNodePolygon.iClampToTextureSize = iClampToTextureSize;
    }
    
    public override void Load(XmlNode aXmlNode) {
        base.Load(aXmlNode);
        
        XmlNodeList list;
        list = aXmlNode.SelectNodes("RenderState");
        if(list != null) {
            Assert.Check(list.Count == 1);
            iRenderState.Load(list[0]);
        }
        
        list = aXmlNode.SelectNodes("ClampToTextureSize");
        if(list != null) {
            if(list.Count > 0) {
                Assert.Check(list.Count == 1);
                iClampToTextureSize = bool.Parse(list[0].FirstChild.Value);
            }
        }
    }
    
    public override void Link() {
        //Trace.WriteLine(Trace.kGui, "Linking NodePolygon: " + Fullname);
        base.Link();
        iRenderState.Link();
        if(iRenderState.Texture.Object != null && iClampToTextureSize == true) {
            if(iRenderState.Texture.Object.Loaded()) {
                iWidth = iRenderState.Texture.Object.Width;
                iHeight = iRenderState.Texture.Object.Height;
            }
        }
    }
    
    public override void Save(XmlTextWriter aWriter) {
        base.Save(aWriter);
        
        iRenderState.Save(aWriter);
        
        aWriter.WriteStartElement("ClampToTextureSize");
        aWriter.WriteString(iClampToTextureSize.ToString());
        aWriter.WriteEndElement();
    }
    
    public override void Draw(IRenderer aRenderer) {
        /*if(iRenderState.Texture.Object != null) {
            if(iRenderState.Texture.Object.Width > 0 && iRenderState.Texture.Object.Height > 0) {
                if(iRenderState.Texture.Object.Width != iWidth || iRenderState.Texture.Object.Height != iHeight) {
                    iWidth = iRenderState.Texture.Object.Width;
                    iHeight = iRenderState.Texture.Object.Height;
                    ObserverUpdate();
                }
            }
        }*/
        aRenderer.DrawPolygon(this);
    }
    
    public override void Visit(Visitor aVisitor, bool aVisitNonActive) {
        aVisitor.AcceptPolygon(this);
        base.Visit(aVisitor, aVisitNonActive);
    }
    
    public override void Texture(ReferenceTexture aTexture) {
        iRenderState.Texture = aTexture;
        if(iRenderState.Texture.Object != null && iClampToTextureSize == true) {
            if(iRenderState.Texture.Object.Loaded()) {
                iWidth = iRenderState.Texture.Object.Width;
                iHeight = iRenderState.Texture.Object.Height;
                ObserverUpdate();
            }
        }
    }
    
    public RenderState CurrRenderState {
        get {
            return iRenderState;
        }
    }
    
    public bool ClampToTextureSize {
        get {
            return iClampToTextureSize;
        }
        set {
            iClampToTextureSize = value;
            if(iRenderState.Texture.Object != null && iClampToTextureSize == true) {
                iWidth = iRenderState.Texture.Object.Width;
                iHeight = iRenderState.Texture.Object.Height;
            }
            ObserverUpdate();
        }
    }
    
    public override void OnSetWidth() {
        base.OnSetWidth();
        if(iRenderState.Texture.Object != null && iClampToTextureSize) {
            iWidth = iRenderState.Texture.Object.Width;
        }
    }
    
    public override void OnSetHeight() {
        base.OnSetHeight();
        if(iRenderState.Texture.Object != null && iClampToTextureSize) {
            iHeight = iRenderState.Texture.Object.Height;
        }
    }
    
    private RenderState iRenderState = new RenderState();
    private bool iClampToTextureSize = true;
}

} // Scenegraph
} // Gui
} // Linn
